src/charm.py

```diff
@@ -74,31 +77,47 @@ class GitSyncLayer(LayerBuilder):
     # Since this is an implementation detail, it is captured here as a class variable.
     SUBDIR: Final = "repo"
 
-    def __init__(self, service_name: str, repo: str, branch: str, wait: int, root: str):
+    def __init__(
+        self,
+        service_name: str,
+        *,
+        repo: str,
+        branch: str,
+        rev: str,
+        wait: int,
+        root: str,
+        port: int,
+    ):
         super().__init__(service_name)
         if not repo:
             raise LayerConfigError("git-sync config error: invalid repo")
-        elif not branch:
-            raise LayerConfigError("git-sync config error: invalid branch")
-        elif wait <= 0:
-            raise LayerConfigError("git-sync config error: wait time must be > 0")
 
         self.repo = repo
         self.branch = branch
+        self.rev = rev
         self.wait = wait
         self.root = root
+        self.port = port
 
     def _command(self) -> str:
-        cmd = (
-            "/git-sync "
-            f"-repo {self.repo} "
-            f"-branch {self.branch} "
-            "-depth 1 "
-            f"-wait {self.wait} "
-            # "-git-config k:v,k2:v2 "
-            f"-root {self.root} "
-            f"-dest {self.SUBDIR}"  # so charm code doesn't need to delete
-        )
+        args = [
+            "/git-sync",
+            f"-repo {self.repo}",
+            "-depth 1",
+            f"-root {self.root}",
+            f"-dest {self.SUBDIR}",  # so charm code doesn't need to delete
+            f"--http-bind :{self.port}",
+            "--http-metrics true",
+        ]
+
+        if self.branch:
+            args.append(f"-branch {self.branch}")
+        if self.rev:
+            args.append(f"-rev {self.rev}")
```

Comment:
Wow this design with separate options for branches and other references is… bad of git-sync :-(

Reply:
I am wondering if using `/refs/heads/<branch>` works when passing it as a reference to git-sync. @sed-i let’s pole around a bit, I’d much rather have just one config option to the charm, rather than expose the questionable design of git-sync to the Juju admin.

Reply:
See reply above.